### DISCUSSION TITLE
Only validate W3C capabilities when `alwaysMatch` is present

### DISCUSSION THREAD

**Discussion by lukyth** [webdriverio/webdriverio#6570] [Technical Discussion]
https://github.com/webdriverio/webdriverio/blob/e9b565e0cb6aeca69d331cbd883734ec72af04f1/packages/webdriver/src/utils.ts#L28-L48

Right now it's validating that the capabilities are W3C compliance when there are some extension capabilities present.

```js
if (extensionCaps.length && invalidWebDriverCaps.length)
```

But there could be a use case that still unable to migrate away from JSONWP capabilities yet and need to have extension capabilities in there at the same time for it to work. For example, only `ms:edgeOptions` will work and `edgeOptions` won't. So using extension capabilities as a signal for W3C capabilities might give some false positives.

So what I'm purposing is, how about the validation only trigger when `alwaysMatch` is present in the capabilities? That way it's guaranteed that the capabilities are _intended_ to be W3C compliant. This also aligns with the code following the validation that uses `alwaysMatch` as a way to tell whether the capabilities are W3C compliance or not.

```js
params.capabilities && (params.capabilities as Capabilities.W3CCapabilities).alwaysMatch
  ?
  :
```

https://github.com/webdriverio/webdriverio/blob/e9b565e0cb6aeca69d331cbd883734ec72af04f1/packages/webdriver/src/utils.ts#L56-L64

So this is what I'm thinking about:

```ts
function isW3CCaps(caps: Capabilities.DesiredCapabilities | Capabilities.W3CCapabilities): caps is Capabilities.W3CCapabilities {
    return Boolean(caps && (caps as Capabilities.W3CCapabilities).alwaysMatch)
}
///
///
if (isW3CCaps(params.capabilities)) {
    const alwaysMatchCaps = Object.keys(params.capabilities.alwaysMatch)
    const firstMatchCaps = params.capabilities.firstMatch
        ? params.capabilities.firstMatch
            .map(caps => Object.keys(caps))
            .reduce((pv, cv) => pv.concat(cv), [])
        : []

    const allCaps = [...new Set(alwaysMatchCaps.concat(firstMatchCaps))]

    const invalidWebDriverCaps = allCaps.filter((cap) => !VALID_CAPS.includes(cap) && !cap.includes(':'))

    if (invalidWebDriverCaps.length) {
        throw new Error(
            `Invalid or unsupported WebDriver capabilities found ("${invalidWebDriverCaps.join('", "')}"). ` +
            'Ensure to only use valid W3C WebDriver capabilities (see https://w3c.github.io/webdriver/#capabilities).'
        )
    }
}
```

**Comment by christian-bromann** [comment]
> But there could be a use case that still unable to migrate away from JSONWP capabilities yet

Can you explain why that is the case? Using `alwaysMatch/firstMatch` capabilities has not become very wide know and popular given it is not a common desired behavior (most of the people want the capabilities they define and not a variation of it). WebdriverIO therefor automatically transforms the capabilities into `alwaysMatch/firstMatch`.

**Comment by lukyth** [comment]
> Can you explain why that is the case?

I was trying to migrate some of the capabilities I use at the company before this, but I couldn't make them work at the time. After spending more time on it, I've found out now that some of them were only partially matched before, so they couldn't match when I was migrating and I actually have to remove some capabilities to make them able to match 🤦 😅 

I understand the idea behind not relying on users passing `alwaysMatch/firstMatch` now. Thank you so much!
